### PR TITLE
feature: updates

### PR DESCRIPTION
**Pull request by domharrington**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -29,67 +28,64 @@ DatabaseUpdates.prototype = Object.create(EventEmitter.prototype)
...
   try {
-    this.updateFiles = fs.readdirSync(this.updatePath)
+    this.updateFiles = fs
+      .readdirSync(this.updatePath)
       // exclude non-javascript files in the updates folder
-      .map(i => (path.extname(i) !== '.js' ? false : i))
+      .map((i) => (path.extname(i) !== '.js' ? false : i))
       // exclude falsy values and filenames that without a valid semver
-      .filter(i => i && semver.valid(i.split('-')[0]))
+      .filter((i) => i && semver.valid(i.split('-')[0]))
```
Comment (cvan) [inline index.js:36]:
```suggestion
      .filter((fn) => ['.js', '.ts'].includes(path.extname(fn)) && semver.valid(fn.split('-')[0]))
```

File: index.js [hunk 2]
```diff
@@ -29,67 +28,64 @@ DatabaseUpdates.prototype = Object.create(EventEmitter.prototype)
 
 DatabaseUpdates.prototype.getFiles = function getFiles() {
   try {
-    this.updateFiles = fs.readdirSync(this.updatePath)
+    this.updateFiles = fs
+      .readdirSync(this.updatePath)
       // exclude non-javascript files in the updates folder
-      .map(i => (path.extname(i) !== '.js' ? false : i))
+      .map((i) => (path.extname(i) !== '.js' ? false : i))
```
Comment (cvan) [inline index.js:34]:
```suggestion
```

File: index.js [hunk 3]
```diff
@@ -29,67 +28,64 @@ DatabaseUpdates.prototype = Object.create(EventEmitter.prototype)
 
 DatabaseUpdates.prototype.getFiles = function getFiles() {
   try {
-    this.updateFiles = fs.readdirSync(this.updatePath)
+    this.updateFiles = fs
+      .readdirSync(this.updatePath)
       // exclude non-javascript files in the updates folder
```
Comment (cvan) [inline index.js:33]:
```suggestion
      // exclude non-javascript/typescript files in the updates folder, and
```

File: index.js [hunk 4]
```diff
@@ -29,67 +28,64 @@ DatabaseUpdates.prototype = Object.create(EventEmitter.prototype)
...
-      .map(i => (path.extname(i) !== '.js' ? false : i))
+      .map((i) => (path.extname(i) !== '.js' ? false : i))
       // exclude falsy values and filenames that without a valid semver
```
Comment (cvan) [inline index.js]:
```suggestion
      // exclude falsy values and filenames without a valid semver
```

File: index.js [hunk 5]
```diff
@@ -29,67 +28,64 @@ DatabaseUpdates.prototype = Object.create(EventEmitter.prototype)
 
 DatabaseUpdates.prototype.getFiles = function getFiles() {
   try {
-    this.updateFiles = fs.readdirSync(this.updatePath)
+    this.updateFiles = fs
```
Comment (cvan) [inline index.js:31]: curious why semver major version dictates the list of pending "data migrations" - versus a `applied` boolean (or `dateApplied` timestamp. or, the inverse, an enum: `"pending" | "success" | "error"`.

that's the pattern I'm familiar with, but the current behaviour does make logical; breaking changes necessitate a major semver upgrade, which necessitates running a new migration file.

knowing which ones have been applied, and which ones have yet to be applied, feels safer from user error. I have to remember to bump the package every time a migration is added. accidents happen. without the guardrails, a migration could mistakingly not be run, whilst code is deployed expecting the schema/data update to have been applied.

to add another layer of paranoid caution: data migrations and human mistakes can go hand in hand. so, beyond storing the `dateApplied` timestamp, storing a hash of the file content (like [`migrate-mongo` does](https://github.com/seppevs/migrate-mongo/blob/master/lib/actions/up.js#L38-L44)) - not just the filename - could prove useful to those botched the migration, bumped the package version, and shipped it to production.

take all this with a grain of salt. I'm relaying this only from past incidents I've encountered. with carefulness, none of this should be necessary.

File: index.js [hunk 6]
```diff
@@ -29,67 +28,64 @@ DatabaseUpdates.prototype = Object.create(EventEmitter.prototype)
...
 }
 
-DatabaseUpdates.prototype.run = function run() {
-  async.eachSeries(this.updateFiles, this.runFile.bind(this), (err) => {
-    if (err) return this.logger.error('Error running updates:', err)
+DatabaseUpdates.prototype.run = async function run() {
+  try {
+    // eslint-disable-next-line no-restricted-syntax
+    for (const file of this.updateFiles) {
+      // eslint-disable-next-line no-await-in-loop
```
Comment (cvan) [inline index.js:81]: don't want to build an array and use `Promise.all` or `Promise.allSettled`? heh

File: .gitignore [hunk 7]
```diff
@@ -1,2 +1,3 @@
 node_modules
 coverage
+.nyc_output
```
Comment (cvan) [inline .gitignore:3]: no *.log or .DS_Store?! (jk)

File: .prettierrc.json [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{
+  "semi": false,
```
Comment (cvan) [inline .prettierrc.json:2]: what kind of world do we live in?

ASI doesn't exist!!;

File: index.js [hunk 9]
```diff
@@ -29,67 +28,64 @@ DatabaseUpdates.prototype = Object.create(EventEmitter.prototype)
...
-      }
+  /* eslint-disable global-require, import/no-dynamic-require */
+  const update = require(path.join(this.updatePath, file))
+  /* eslint-enable global-require, import/no-dynamic-require */
 
-      return this.persistUpdate(file, next)
+  return update(this.db)
+    .then(() => this.persistUpdate(file))
+    .catch((err) => {
+      this.logger.error('Error running update:', file)
```
Comment (cvan) [inline index.js:62]: I follow the pattern, but curious why not emit an 'error'?

Comment (cvan) [reply index.js:62]: no rollback/revert? I don't know about them, but MongoDB transactions are a thing, right?

say the creation of the migration worked, but running it failed. unless I'm misunderstanding the code, I don't see that failure persisted to the DB (or the DB document deleted) - so the library can attempt to re-run it or skip it in the future.

**Review by cvan — APPROVED** [review]
nice library! it's a really useful utility, and concise.

File: .prettierrc.json (hunk 8 above)
Comment (domharrington) [reply .prettierrc.json:2]: Haha i wrote this in 2015 when no semicolons was the way to write JS. I'd like to preserve some of that history :D

File: index.js (hunk 1 above)
Comment (domharrington) [reply index.js:36]: I dont think there's any way for this module to reasonably accept TS files without bringing in typescript as a dependency to allow us to import TS code and call it dynamically, unfortunately.

File: index.js (hunk 6 above)
Comment (domharrington) [reply index.js:81]: Promise.all executes everything at once, whereas we want to apply the files sequentially in the order we find them.

File: index.js (hunk 5 above)
Comment (domharrington) [reply index.js:31]: The file hash idea is a good one! If we wanna use this, i'm happy to build it out, or PRs welcome.

File: index.js (hunk 9 above)
Comment (domharrington) [reply index.js:62]: Yeah I didnt build in rollbacks or reverts when I first wrote this.

File: index.js (hunk 1 above)
Comment (cvan) [reply index.js:36]: my suggestion was more to drop the map above and just use a filter. but I was just nit'ing.

File: index.js (hunk 6 above)
Comment (cvan) [reply index.js:81]: oh, yeah. I forgot that. good call then.

**Pull request merged**